[TO BE REVIEWED]v3.5.0-sc core-zone updates (#2250)
vesoft-inc/nebula-docs: docs-2.0/4.deployment-and-installation/5.zone.md

# Manage Zones

!!! enterpriseonly

    This feature is only available in the NebulaGraph Enterprise Edition.

The Zone is a logical rack of storage nodes in a NebulaGraph cluster. It divides multiple storage nodes into manageable logical areas to achieve resource isolation. At the same time, you can control the Graph service to access the replica data in the specified Zone to reduce traffic consumption and improve access efficiency. This article describes how to use the Zone feature.

## Principle

Within NebulaGraph, you can set up multiple Zones, with each Zone containing one or more Storage nodes. When creating a graph space, the system automatically recognizes these set Zones and stores graph space data on the Storage nodes within these Zones.

> It's important to note that when creating a graph space, you need to specify the number of replicas for data partitioning. At this point, **the specified number of partition replicas and the number of set Zones must be equal, otherwise, the graph space cannot be created**. This is because NebulaGraph evenly stores each replica of the graph space data across these set Zones.

Reviewer: Suggest change to be more precise:

`This is because NebulaGraph evenly distributes the replicas of a graph space partition across these set Zones.`

Since each Zone contains a complete set of graph space data partitions, at least one Storage node is required within each Zone to store these data partitions.

The partition replicas in NebulaGraph achieve strong consistency through the [Raft](../1.introduction/3.nebula-graph-architecture/4.storage-service.md#raft_1) protocol. It's recommended to use an odd number of partition replicas, and therefore, it's also suggested to set an odd number of Zones.

> Taking the following picture as an example, when creating a graph space, the data is partitioned into 3 partitions, with 3 replicas for each partition, and the number of Zones is also 3. Six machines starting the Storage service are paired up and added to these 3 Zones. When creating the graph space, NebulaGraph evenly stores each replica of the graph space data across zone1 to zone3, and each Zone contains a complete set of graph space data partitions (part_1 to part_3).

Reviewer: `Six machines starting the Storage ...` -> `Six machines hosting the Storage service...`

![Zone](https://docs-cdn.nebula-graph.com.cn/figures/zone1.png)

> To reduce traffic cost and bandwidth consumption of cross-Zone queries, you can configure the Graph service to access a specific single Zone. When the Graph service accesses a specific Zone, the Graph service will access the replica data in the specified Zone.

Reviewer: `To reduce cost of cross-Zone network traffic, and increase data transfer speed (i.e., intra-zone network usually has a lower latency than inter-zone network), you can configure the Graph service to prioritize intra zone data access. Each Graphd will then prioritize to access the partition replica in the same zone if there is any.`

## Scenarios

- Resource isolation. You can create a graph space on specified storage nodes to achieve resource isolation.
- Rolling upgrade. You need to stop one or more servers to update them, and then put them into use again until all servers in the cluster are updated to the new version.
- Cost saving. Allocate different graph spaces to different Zones, and control the client to access the replica data in the specified Zone to reduce traffic consumption and improve access efficiency.

## Notes

- Before enabling the Zone feature, clear any existing data in the cluster. See **Enabling Zone** for details.
- Each Storage node must belong to one, and only one, Zone. However, a Zone can have multiple Storage nodes. Storage nodes should outnumber or equal Zones.

> - The number of Zones must equal the number of partition replicas; otherwise, the graph space cannot be created.

Reviewer: turn it around

# of replicas must equal # zones

> - The number of Zones is recommended to be odd and must equal the number of partition replicas.

Reviewer: ditto

- Adjusting the number of Zones isn't allowed.
- Zone name modifications are unsupported.

## Enable Zone 

> 1. In the configuration file `nebula-metad.conf` of the Meta service, set `--zone_list` to Zone names to be added, such as `--assigned_zone=zone1, zone2, zone3`.

Reviewer: assigned_zone?

  !!! danger

        Once the value of `--zone_list` is configured and the Meta service is started, it cannot be modified, otherwise, the Meta service will fail to restart. 

  !!! note

      - The number of Zones specified in `--zone_list` is recommended to be odd and must be less than or equal to the number of Storage nodes. When `--zone_list` is empty, it indicates that the Zone feature is disabled.
      - Consider the replica settings when setting the number of Zones, since the number of Zones should match the replica count. For example, with 3 replicas, you must have 3 Zones.

>       - If the name of a Zone contains special characters (excluding underscores), reserved keywords, or starts with a number, you need to enclose the Zone name in backticks (`) when specifying the Zone name in a query statement; the Zone name cannot contain English periods (.); multiple Zone names are separated by commas (,).

Reviewer: `multiple Zone names must be separated by commas (,)`

<img src="https://docs-cdn.nebula-graph.com.cn/figures/core_zone_230913.png" width="800" alt="example_for_zones">

To reduce cost of cross-Zone network traffic, and increase data transfer speed (Intra-zone network usually has a lower latency than inter-zone network), you can configure the Graph service to prioritize intra zone data access. Each Graphd will then prioritize to access the partition replica in the same zone as specified by Graphd if there is any. As an example, suppose Graph A and Graph B are located in zone1, Graph C and Graph D are in zone2, and Graph E is in zone3. You can configure Graph A and Graph B to prioritize accessing data in zone1, Graph C and Graph D to prioritize accessing data in zone2, and Graph E to prioritize accessing data in zone3. This helps reduce the cost of cross-zone network traffic and improves data transfer speed.

<img src="https://docs-cdn.nebula-graph.com.cn/figures/core_intra_zone_20230913.png" width="800" alt="example_for_intra_zone">

> In cases where data retrieval fails due to a Storage node failure in a specific zone, you can configure the Graph service to read leader replica data across zones to ensure data availability. For instance, if machine B experiences a failure, causing Graph A to be unable to read the Follower replica data of Part3 in zone1, you can configure Graph A to read the Leader replica data of Part3 in zone3, thus ensuring data availability.

Reviewer: Hold  on a second. This may not be the case at this moment. Let me confirm.

Reviewer: Let's delete this paragraph. It's the future feature. Currently on storaged failure, we will get 'E_RPC_FAILURE' from the response of a storaged request. Then we will either return this code or ignore the code if `accept_partial_success` is set to true. In the future, after we implement the retry functionality, we will have ways to retry request to other replicas (not necessarily leader replica). In the current implementation, we do not do any retry. We just determine whether send request to leader or the replica in the same zone in advance. We only send request to leader in the following cases:
```
When intra zone reading is enabled, if zone is not given or the zone is empty, or there is no part host in this zone (impossible under the constraint of #replicas = # zones) && allow rerouting, fallback to leader. If intra zone reading is
disabled, fall back to leader. If it's write request, fall back to leader.
```

<img src="https://docs-cdn.nebula-graph.com.cn/figures/core_zone_230913.png" width="800" alt="example_for_zones">

To reduce cost of cross-Zone network traffic, and increase data transfer speed (Intra-zone network usually has a lower latency than inter-zone network), you can configure the Graph service to prioritize intra-zone data access. Each Graphd will then prioritize to access the partition replica in the same zone as specified by Graphd if there is any. As an example, suppose Graph A and Graph B are located in zone1, Graph C and Graph D are in zone2, and Graph E is in zone3. You can configure Graph A and Graph B to prioritize accessing data in zone1, Graph C and Graph D to prioritize accessing data in zone2, and Graph E to prioritize accessing data in zone3. This helps reduce the cost of cross-zone network traffic and improves data transfer speed.

<img src="https://docs-cdn.nebula-graph.com.cn/figures/core_intra_zone_20230913.png" width="800" alt="example_for_intra_zone">

<!-- To be released next version 

> In cases where data retrieval fails due to a Storage node failure in a specific Zone, you can configure the Graph service to read leader replica data across zones to ensure data availability. For instance, if machine B experiences a failure, causing Graph A to be unable to read the Follower replica data of Part3 in zone1, you can configure Graph A to read the Leader replica data of Part3 in zone3, thus ensuring data availability.

Author: > 如果当前zone就是leader，且返回失败了呢，还能出去么？

这部分内容隐藏了，这个功能还咩有实现。

   
  For more information about the Meta configuration file, see [Meta service configuration](../5.configurations-and-logs/1.configurations/2.meta-config.md).

2. Restart the Meta service.

## Specify intra Zone data access

1. Enable the Zone feature. For details, see **Enable Zone** above.
2. In the configuration file `nebula-graphd.conf` of the Graph service, add the following configuration:
  
  1. Set the `--assigned_zone` to the name of the Zone where the Graphd is assigned, such as `--assigned_zone=zone1`.
   
    !!! note
      
        - Different Graph services can set different values for `--assigned_zone`, but the value of `--assigned_zone` must be one of the values in `--zone_list`. In production, it is recommended to use the actual zone that a Graphd locates to reduce management complexity. Of course, it must be within the `zone_list`. Otherwise, intra zone reading may not take effect.
        - The value of `--assigned_zone` is a string and does not support English commas (,).
        - When `--assigned_zone` is empty, it means reading from leader replicas.
  

>   2. Set `--prioritize_intra_zone_reading` to `true` to prioritize intra zone data reading. When there is no data in the Zone specified by `--assigned_zone`, an error occurs depending on the value of `stick_to_intra_zone_on_failure`. 

Author: ```suggestion
  2. Set `--prioritize_intra_zone_reading` to `true` to prioritize intra zone data reading. When reading fails in the Zone specified by `--assigned_zone`, an error occurs depending on the value of `stick_to_intra_zone_on_failure`. 
```

1. Enable the Zone feature. For details, see **Enable Zone** above.
2. In the configuration file `nebula-graphd.conf` of the Graph service, add the following configuration:
  

>   1. Set the `--assigned_zone` to the name of the Zone to be accessed, such as `--assigned_zone=zone1`.

Reviewer: 这里assigned zone是指这台graphd被schedule/assign的zone，也就是说这台graphd实际所在的zone。

   
    !!! note
      

>         - Different Graph services can set different values for `--assigned_zone`, but the value of `--assigned_zone` must be one of the values in `--zone_list`.

Reviewer: 不用。graphd所在的zone不需要与storaged一样。导致的后果就是需要prioritize intra zone reading的时候，找不到，最后就从leader中读。所以逻辑是：
1. assigned_zone就是这个graphd真实所在的zone
2. If you want to make sure each graphd only read data in the same zone, make sure the assigne_zone is with the list of zone_list.

>         - Different Graph services can set different values for `--assigned_zone`, but the value of `--assigned_zone` must be one of the values in `--zone_list`.

Reviewer: 最好加一句：In production, it is suggested to use the actual zone that a graphd locates to reduce management complexity.

        - The value of `--assigned_zone` is a string and does not support English commas (,).

>         - When the value of `--assigned_zone` is empty, it indicates that all Zones are accessed.

Reviewer: It means read from leaders.

| Hosts           | Port |
+-----------------+------+
| "192.168.8.111" | 7779 |
| "192.168.8.112" | 9779 |
+-----------------+------+
```

### Create a space in the specified Zones

The syntax for creating a graph space within a Zone is the same as in [Creating Graph Space](../3.ngql-guide/9.space-statements/1.create-space.md). 

> However, during graph space creation, the system automatically recognizes the `--zone_list` value from the Meta configuration file. If this value is not empty and the number of Zones matches the partition replica count specified by `replica_factor`, the graph space's replicas will be evenly distributed across the Zones in `--zone_list`. If the specified replica count doesn't match the number of Zones, graph space creation will fail.

Reviewer: 这里的value是指 [ON <zone_list>]里面的zone_list吧？如果是，请看下面修改。

`If this value is not empty and it is the same as the value specified in \`--zone_list\` from the Meta configuration file .... Otherwise, graph space creation will fail.
